Created the load script for the pipeline | Improved extract and transform script

CharliedeanCS opened the pull request:
Worked on 3 files:

load.py
Created a load script that takes in the most recent collection of plant data and uploads it to the SQL Server
- Pylinted the code and improved efficieny

transform.py
Changed the transform file so it interacts with the data folder as this is where the data will be located

extract.py
Changed the extract script so it now extracts the plant id as inserts it into the database and ensures to keep the plant id from the API instead of creating its own index.

CharliedeanCS committed:
Added a database connection

CharliedeanCS committed:
added insert into tables function

CharliedeanCS committed:
seperated csv making into functions

CharliedeanCS committed:
added csv conversion functions

CharliedeanCS committed:
Merge branch 'main' of https://github.com/CharliedeanCS/Ladybirds-Plant-Sensors-Project into create-load-script

CharliedeanCS committed:
Merge branch 'main' of https://github.com/CharliedeanCS/Ladybirds-Plant-Sensors-Project into create-load-script

CharliedeanCS committed:
added insert function into SQL Server

CharliedeanCS committed:
added a plant id column back to the csv

CharliedeanCS committed:
Fixed load script so it includes ID numbers

CharliedeanCS committed:
Improved Code quality : extract script line length

zandersnow14 commented on pipeline/load.py:
```diff
@@ -0,0 +1,59 @@
+"""Script to connect to an MSSQL database and insert plant data"""
```
```suggestion
"""Script to connect to an MSSQL database and insert plant data."""
```

zandersnow14 commented on pipeline/load.py:
```diff
@@ -0,0 +1,59 @@
+
+
+def create_database_connection() -> Connection:
+    """Creates a database connection to the SQL Server"""
```
```suggestion
    """Creates a database connection to the SQL Server."""
```

zandersnow14 commented on pipeline/load.py:
```diff
@@ -0,0 +1,59 @@
+
+
+def insert_into_recordings_table(connection: Connection, plant_data: pd.DataFrame) -> None:
+    """Inserts data from a pandas data frame into a SQL Server"""
```
```suggestion
    """Inserts data from a pandas data frame into a SQL Server."""
```

zandersnow14 commented on pipeline/load.py:
```diff
@@ -0,0 +1,59 @@
+                VALUES (:id,:soil,:temperature,:recording,:watered)""")
+            connection.execute(query, {"id": id, "soil": soil, "temperature": temperature, "recording": recording,
+                                       "watered": watered})
+        except:
```
Could be more specific with the exception here - what might go wrong here to require an exception?

CharliedeanCS replied:
I decided to remove the try-except clause as I realised it isn't needed and never runs into a problem at this part

shivani-63 commented on pipeline/load.py:
```diff
@@ -0,0 +1,59 @@
+            connection.execute(query, {"id": id, "soil": soil, "temperature": temperature, "recording": recording,
+                                       "watered": watered})
+        except:
+            pass
```
does this just leave values as null if they cannot be located using your try/except clause?

zandersnow14 commented on pipeline/transform.py:
```diff
@@ -66,26 +66,38 @@ def change_temp_and_moisture_to_two_dp(dataframe: pd.DataFrame) -> pd.DataFrame:
+
+
+def upload_clean_csv_file(filename: str, plant_data: pd.DataFrame) -> None:
+    """Creates a new CSV file with the clean data"""
```
```suggestion
    """Creates a new CSV file with the clean data."""
```

zandersnow14 commented on pipeline/load.py:
```diff
@@ -0,0 +1,59 @@
+from transform import csv_to_data_frame
+
+
+def create_database_connection() -> Connection:
```
Quite nit picky, but maybe think about passing in a `config` argument here (so replace `environ` in this function with `config`, and then pass in environ in the `if __name__ == "__main__"` statement - makes things a bit neater in my head but just a thought.

shivani-63 commented on pipeline/transform.py:
```diff
@@ -66,26 +66,38 @@ def change_temp_and_moisture_to_two_dp(dataframe: pd.DataFrame) -> pd.DataFrame:
-    df = check_soil_temp_valid(df)
-    df = normalize_datetimes(df)
-    df = change_temp_and_moisture_to_two_dp(df)
+    plants = check_soil_moisture_valid(plants)
```
Really nice changes to variable names here, makes things clear and relevant!

CharliedeanCS committed:
removed try and except | added config parameter

zandersnow14 commented on pipeline/load.py:
```diff
@@ -0,0 +1,56 @@
+"""Script to connect to an MSSQL database and insert plant data"""
+
+
+from os import environ
```
```suggestion
from os import environ, _Environ
```

zandersnow14 commented on pipeline/load.py:
```diff
@@ -0,0 +1,56 @@
+from transform import csv_to_data_frame
+
+
+def create_database_connection(config) -> Connection:
```
```suggestion
def create_database_connection(config: _Environ) -> Connection:
```

CharliedeanCS committed:
Added type hint for config

Co-authored-by: Zander-Snow-Sigma <145514262+Zander-Snow-Sigma@users.noreply.github.com>

zandersnow14 committed:
Add _Environ import

Added _Environ import to be used for type hinting config variable in the get_db_conn function.
